flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -365,4 +401,18 @@ public String toString() {
             return "EmptyDiscardStateObject{" + stateHandleID + '}';
         }
     }
+
+    private boolean preventsDiscardingCreatedCheckpoint(SharedStateEntry entry) {
+        // explicitly set by the backend, e.g. private state is reused
+        if (entry.preventDiscardingCreatedCheckpoint) {
+            return true;
+        }
+        // with NO_SHARING strategy, shared state, if any, is bundled inside the checkpoint folder
+        // noinspection RedundantIfStatement
```

Comment:
same here, add comments saying NO_SHARING indicates recover from a savepoints, and explain why delay deletion is needd.